File: src/cs/PhantasyIslandPythonRemoteControl/UAV.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;

namespace PhantasyIslandPythonRemoteControl
{
    public class AirplaneController : AirplaneCore
    {
        private int _count = 1;
        private Func<string, Task<JsonElement?>> _sendCmdFn;

        public AirplaneController() : base()
        {
            _sendCmdFn = HttpLayer.SendCmd;
        }

        public void UseFastMode(bool fastMode = true, bool asyncMode = true)
        {
            if (asyncMode)
            {
                _sendCmdFn = fastMode ? HttpLayer.SendCmdVolatile : HttpLayer.SendCmd;
            }
            else
            {
                // Synchronous wrapper (basic implementation)
                _sendCmdFn = s => Task.Run(() => (fastMode ? HttpLayer.SendCmdVolatile(s) : HttpLayer.SendCmd(s)).Result);
            }
        }

        private string NextCount()
        {
            _count += 2;
            return _count.ToString();
        }

        private string PrepareCommand(string command) => $"{KeyName} {NextCount()} {command}";

        private async Task<JsonElement?> SendCommand(string command)
        {
            return await _sendCmdFn(PrepareCommand(command));
        }

        public Task Mode(int mode) => AirplaneMode(mode);
        public Task Takeoff(int high) => SendCommand($"takeoff {high}");
        public Task Land() => SendCommand("land");
        public Task Emergency() => SendCommand("emergency");
        public Task Up(int distance) => SendCommand($"up {distance}");
        public Task Down(int distance) => SendCommand($"down {distance}");
        public Task Forward(int distance) => SendCommand($"forward {distance}");
        public Task Back(int distance) => SendCommand($"back {distance}");
        public Task Left(int distance) => SendCommand($"left {distance}");
        public Task Right(int distance) => SendCommand($"right {distance}");
        public Task Goto(int x, int y, int h) => SendCommand($"goto {x} {y} {h}");
        public Task Flip(string direction) => SendCommand($"flip {direction} 1");
        public Task FlipForward() => Flip("f");
        public Task FlipBack() => Flip("b");
        public Task FlipLeft() => Flip("l");
        public Task FlipRight() => Flip("r");
        public Task Rotate(int degree) => SendCommand($"rotate {degree}");
        public Task Cw(int degree) => SendCommand($"cw {degree}");
        public Task Ccw(int degree) => SendCommand($"ccw {degree}");
        public Task High(int high) => SendCommand($"high {high}");
        public Task Speed(int speed) => SendCommand($"setSpeed {speed}");
        public Task Led(int r, int g, int b) => SendCommand($"light {r} {g} {b}");
        public Task Bln(int r, int g, int b) => SendCommand($"bln {r} {g} {b}");
        public Task Rainbow(int r, int g, int b) => SendCommand($"rainbow {r} {g} {b}");
        public Task AirplaneMode(int mode) => SendCommand($"airplane_mode {mode}");
        public Task Stop() => Hover();
        public Task Hover() => SendCommand("hover");
    }

    public class AirplaneManager
    {
        public Dictionary<string, AirplaneController> AirplanesTable { get; } = new Dictionary<string, AirplaneController>();

        public async Task<JsonElement?> Ping() => await HttpLayer.Ping();
        public async Task<JsonElement?> Start() => await HttpLayer.Start();

        public AirplaneController GetAirplane(string id)
        {
            return AirplanesTable.TryGetValue(id, out var air) ? air : null;
        }

        public async Task Flush()
        {
            var statusJson = await HttpLayer.GetAllAirplaneStatus();
            if (statusJson == null) return;
            
            var airplaneStatus = HttpLayer.ProcessAirplane(statusJson.Value);
            if (airplaneStatus == null) return;

            foreach (var kv in airplaneStatus)
            {
                var statusDict = kv.Value as Dictionary<string, object>;
                if (!AirplanesTable.ContainsKey(kv.Key))
                {
                    var air = new AirplaneController
                    {
                        KeyName = (string)statusDict["keyName"],
                        TypeName = (string)statusDict["typeName"],
                        UpdateTimestamp = (double)statusDict["updateTimestamp"],
                        Status = AirplaneFlyStatus.FromJson((JsonElement)statusDict["status"]),
                        CameraFront = (string)statusDict["cameraFront"],
                        CameraDown = (string)statusDict["cameraDown"]
                    };
                    AirplanesTable[kv.Key] = air;
                }
                else
                {
                    var air = AirplanesTable[kv.Key];
                    air.KeyName = (string)statusDict["keyName"];
                    air.TypeName = (string)statusDict["typeName"];
                    air.UpdateTimestamp = (double)statusDict["updateTimestamp"];
                    air.Status = AirplaneFlyStatus.FromJson((JsonElement)statusDict["status"]);
                    air.CameraFront = (string)statusDict["cameraFront"];
                    air.CameraDown = (string)statusDict["cameraDown"];
                }
            }
        }

        private static AirplaneManager _instance = new AirplaneManager();
        public static AirplaneManager Instance => _instance;
    }

    public class UAV
    {
        private AirplaneManager _airs = AirplaneManager.Instance;

        public UAV()
        {
            _airs.Flush().Wait();
            _airs.Start().Wait();
            _airs.Flush().Wait();
        }

        public void Sleep(int seconds) => Task.Delay(TimeSpan.FromSeconds(seconds)).Wait();
        public void AddUav(string port) => _airs.Flush().Wait();
        private AirplaneController P(string port) => _airs.GetAirplane(port);

        public void Land(string port) => P(port).Land().Wait();
        public void Takeoff(string port, int high) => P(port).Takeoff(high).Wait();
        public void Up(string port, int dist) => P(port).Up(dist).Wait();
        public void Down(string port, int dist) => P(port).Down(dist).Wait();
        public void Forward(string port, int dist) => P(port).Forward(dist).Wait();
        public void Back(string port, int dist) => P(port).Back(dist).Wait();
        public void Left(string port, int dist) => P(port).Left(dist).Wait();
        public void Right(string port, int dist) => P(port).Right(dist).Wait();
        public void Goto(string port, int x, int y, int h) => P(port).Goto(x, y, h).Wait();
        public void Rotate(string port, int deg) => P(port).Rotate(deg).Wait();
        public void Stop(string port) => P(port).Stop().Wait();
        public void Hover(string port) => P(port).Hover().Wait();
    }

    public class FH0A : UAV { }
}

File: src/fs/UAV.fs
namespace PhantasyIslandRemoteControl

open System
open System.Collections.Generic
open System.Text.Json
open System.Threading.Tasks

type AirplaneController() =
    inherit AirplaneCore()
    
    let mutable _count = 1
    let mutable _sendCmdFn: string -> Async<JsonElement option> = HttpLayer.sendCmd

    member this.UseFastMode(?fastMode: bool, ?asyncMode: bool) =
        let fast = defaultArg fastMode true
        let asyncM = defaultArg asyncMode true
        if asyncM then
            _sendCmdFn <- if fast then HttpLayer.sendCmdVolatile else HttpLayer.sendCmd
        else
            // Wrapper to run async as sync if needed, though in F# we prefer async
            _sendCmdFn <- fun s -> 
                let res = if fast then HttpLayer.sendCmdVolatile s else HttpLayer.sendCmd s
                async { return Async.RunSynchronously res }

    member private this.NextCount() =
        _count <- _count + 2
        string _count

    member private this.PrepareCommand(command: string) =
        sprintf "%s %s %s" this.KeyName (this.NextCount()) command

    member private this.SendCommand(command: string) =
        _sendCmdFn (this.PrepareCommand(command))

    member this.Mode(mode: int) = this.AirplaneMode(mode)
    member this.Takeoff(high: int) = this.SendCommand(sprintf "takeoff %d" high)
    member this.Land() = this.SendCommand("land")
    member this.Emergency() = this.SendCommand("emergency")
    member this.Up(distance: int) = this.SendCommand(sprintf "up %d" distance)
    member this.Down(distance: int) = this.SendCommand(sprintf "down %d" distance)
    member this.Forward(distance: int) = this.SendCommand(sprintf "forward %d" distance)
    member this.Back(distance: int) = this.SendCommand(sprintf "back %d" distance)
    member this.Left(distance: int) = this.SendCommand(sprintf "left %d" distance)
    member this.Right(distance: int) = this.SendCommand(sprintf "right %d" distance)
    member this.Goto(x: int, y: int, h: int) = this.SendCommand(sprintf "goto %d %d %d" x y h)
    member this.Flip(direction: string) = this.SendCommand(sprintf "flip %s 1" direction)
    member this.FlipForward() = this.Flip("f")
    member this.FlipBack() = this.Flip("b")
    member this.FlipLeft() = this.Flip("l")
    member this.FlipRight() = this.Flip("r")
    member this.Rotate(degree: int) = this.SendCommand(sprintf "rotate %d" degree)
    member this.Cw(degree: int) = this.SendCommand(sprintf "cw %d" degree)
    member this.Ccw(degree: int) = this.SendCommand(sprintf "ccw %d" degree)
    member this.High(high: int) = this.SendCommand(sprintf "high %d" high)
    member this.Speed(speed: int) = this.SendCommand(sprintf "setSpeed %d" speed)
    member this.Led(r: int, g: int, b: int) = this.SendCommand(sprintf "light %d %d %d" r g b)
    member this.Bln(r: int, g: int, b: int) = this.SendCommand(sprintf "bln %d %d %d" r g b)
    member this.Rainbow(r: int, g: int, b: int) = this.SendCommand(sprintf "rainbow %d %d %d" r g b)
    member this.AirplaneMode(mode: int) = this.SendCommand(sprintf "airplane_mode %d" mode)
    member this.Stop() = this.Hover()
    member this.Hover() = this.SendCommand("hover")

type AirplaneManager private () =
    let _airplanesTable = Dictionary<string, AirplaneController>()

    static let _instance = AirplaneManager()
    static member Instance = _instance

    member this.AirplanesTable = _airplanesTable

    member this.Ping() = HttpLayer.ping()
    member this.Start() = HttpLayer.start()

    member this.GetAirplane(id: string) =
        match _airplanesTable.TryGetValue(id) with
        | (true, air) -> Some air
        | _ -> None

    member this.Flush() =
        async {
            let! statusJson = HttpLayer.getAllAirplaneStatus()
            let airplaneStatus = HttpLayer.processAirplane(statusJson)
            if airplaneStatus <> null then
                for kv in airplaneStatus do
                    let statusDict = kv.Value :?> Dictionary<string, obj>
                    let key = kv.Key
                    let air = 
                        if _airplanesTable.ContainsKey(key) then
                            _airplanesTable.[key]
                        else
                            let a = AirplaneController()
                            _airplanesTable.[key] <- a
                            a
                    
                    air.KeyName <- statusDict.["keyName"] :?> string
                    air.TypeName <- statusDict.["typeName"] :?> string
                    air.UpdateTimestamp <- statusDict.["updateTimestamp"] :?> double
                    air.Status <- Some(AirplaneFlyStatus.FromJson(statusDict.["status"] :?> JsonElement))
                    air.CameraFront <- statusDict.["cameraFront"] :?> string
                    air.CameraDown <- statusDict.["cameraDown"] :?> string
        }

type UAV() =
    let _airs = AirplaneManager.Instance

    do
        _airs.Flush() |> Async.RunSynchronously
        _airs.Start() |> Async.Ignore |> Async.RunSynchronously
        _airs.Flush() |> Async.RunSynchronously

    member this.Sleep(seconds: int) = Task.Delay(TimeSpan.FromSeconds(float seconds)).Wait()
    member this.AddUav(port: string) = _airs.Flush() |> Async.RunSynchronously
    member private this.P(port: string) = _airs.GetAirplane(port)

    member this.Land(port: string) = 
        match this.P(port) with
        | Some air -> air.Land() |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Takeoff(port: string, high: int) = 
        match this.P(port) with
        | Some air -> air.Takeoff(high) |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Up(port: string, dist: int) = 
        match this.P(port) with
        | Some air -> air.Up(dist) |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Down(port: string, dist: int) = 
        match this.P(port) with
        | Some air -> air.Down(dist) |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Forward(port: string, dist: int) = 
        match this.P(port) with
        | Some air -> air.Forward(dist) |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Back(port: string, dist: int) = 
        match this.P(port) with
        | Some air -> air.Back(dist) |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Left(port: string, dist: int) = 
        match this.P(port) with
        | Some air -> air.Left(dist) |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Right(port: string, dist: int) = 
        match this.P(port) with
        | Some air -> air.Right(dist) |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Goto(port: string, x: int, y: int, h: int) = 
        match this.P(port) with
        | Some air -> air.Goto(x, y, h) |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Rotate(port: string, deg: int) = 
        match this.P(port) with
        | Some air -> air.Rotate(deg) |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Stop(port: string) = 
        match this.P(port) with
        | Some air -> air.Stop() |> Async.RunSynchronously |> ignore
        | None -> ()

    member this.Hover(port: string) = 
        match this.P(port) with
        | Some air -> air.Hover() |> Async.RunSynchronously |> ignore
        | None -> ()

type FH0A() =
    inherit UAV()
